### PR TITLE
html attributes

### PR DESCRIPTION
**Pull request by GiggleLiu**
Allow setting HTML attributes as kwargs.

See the test file for details.

### CODE REVIEW

**Comment by GiggleLiu** [comment]
Let me explain this PR with a using case.

rng-sequence

```julia
let
	html1 = md"""
* times (ns)
$(@bind ts TextField((80, 2); placeholder="0.1, 0.2, 0.3...", style="margin-left:30px", id="areatimes"))

* phases (0-1)
$(@bind phis TextField((80, 2); placeholder="0.1, 0.2, 0.3...", style="margin-left:30px", id="areaphases"))
""" |> html
	rb = md"""$(@bind randbutton Button("random"; id="random-generate"))""" |> html
	
	HTML("""
		$html1
		$rb
		<script>
		const randgen = this.querySelector("#random-generate");
		const tsfield = this.querySelector("#areatimes");
		const phisfield = this.querySelector("#areaphases");
		randgen.onclick = (event) => {
			event.preventDefault();
			tsfield.value = Array.from({length: 5}, () => Math.floor((Math.random() * $(tmax-tmin)) + $tmin));
			phisfield.value = Array.from({length: 5}, () => (Math.random().toFixed(3)));
			e = new Event('input');
			tsfield.dispatchEvent(e);
			phisfield.dispatchEvent(e);
		};
		</script>
	""")
end
```

Here, set the `id` of an object is important.

**Comment by fonsp** [comment]
You even added tests 💕

**Comment by lllusion3418** [comment]
I'm not very familiar with Julia, but wouldn't it make more sense to pass a `Dict` / array of `Pair`s instead of using kwargs? This might be better for a few reasons:
* wouldn't it be a potential compatibility problem as it is now? If you wanted to add any new keywords argument, you'd have to worry about whether someone already used it as an html attribute
* any attribute containing `-` (like `data-foo`) is difficult to use with kwargs. As far as I can tell, you'd have to do something like `Slider(1:2; [Symbol("data-foo") =>42]...)`
* attributes like `disabled` could be handled more intuitively: in html you usually write `<input disabled>` instead of `<input disabled="">`, which would be possible using an array but not with kwargs. Boolean arguments (`:disabled => false`) would still be a problem in this case, though, so maybe that should be handled as well?

File: test/runtests.jl [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+    for s in [
+        Slider(1:0.2:3; default=0.4, style="{width=100px;}"),
+        Button(; style="{width=100px;}"),
+        FilePicker(; style="{width=100px;}"),
+        Radio(["A", "B", "C"]; style="{width=100px;}"),
+        Select(["A", "B", "C"]; style="{width=100px;}"),
+        MultiSelect(["A", "B", "C"]; style="{width=100px;}"),
+        CheckBox(; default=true, style="{width=100px;}"),
+        TextField(; style="{width=100px;}"),
+        NumberField(1:0.2:3; default=2, style="{width=100px;}"),
```
Comment (lllusion3418) [inline test/runtests.jl:14]: Why the braces around the `style` values? Quoting [the standard](https://drafts.csswg.org/css-style-attr/#style-attribute) (linked from [here](https://html.spec.whatwg.org/multipage/dom.html#the-style-attribute)):

> The value of the style attribute must match the syntax of the contents of a CSS declaration block (excluding the delimiting braces) [...]

Also wouldn't it be more interesting to test for the exact HTML results instead of just whether or not they're empty?

**Comment by fonsp** [comment]
Good point! Have a look at how Hyperscript solved this problem:
https://github.com/yurivish/Hyperscript.jl/

Maybe we should add it as a dependency! It would make our code a lot more robust.

File: test/runtests.jl (hunk 1 above)
Comment (fonsp) [reply test/runtests.jl:14]: Yes we should not use braces.

**Comment by fonsp** [comment]
In response to https://github.com/fonsp/PlutoUI.jl/pull/39#discussion_r491692867

Ideally we should use https://github.com/JuliaWeb/Gumbo.jl during testing to parse the resulting HTML. But this would not be necessary if we use Hyperscript.jl

**Comment by GiggleLiu** [comment]
> wouldn't it make more sense to pass a Dict / array of Pairs instead of using kwargs?

Agree that pure kwargs design can be flawed.  But here, making a perfect html constructor is not the motivation, saving keyboard typing is.

1. The default constructor is not forbidden, one can still use the default constructor like
```julia
julia> Button("test", Dict("data-foo"=>"50"))
Button("test", Dict("data-foo" => "50"))
```
2. One can always switch back to `html"<input type=button ...>"` when he really need advanced features.

I will add some docstrings and polish the tests to make it more user friendly.

**Comment by lllusion3418** [comment]
> Agree that pure kwargs design can be flawed. But here, making a perfect html constructor is not the motivation, saving keyboard typing is.

Yeah, wasn't sure whether it'd be worth it to give up on kwargs since they _are_ more convenient. It's not that much more to type, though, is it?

> The default constructor is not forbidden, one can still use the default constructor

Yeah, thanks for the reminder, I missed that possibility.

> One can always switch back to html"<input type=button ...>" when he really needs advanced features.

Though especially in those cases a wrapper might be convenient because who wants to deal with e.g. manual html escaping.

> does this make sense?

I'm fine keeping it the way it is, not sure which option is actually better.

---

Another thing: for the cases where `withtag` isn't used (`NumberField`, `Button`, `TextField`, ...; i.e. anywhere you're using your `mkattr()`), wouldn't it be cleaner and easier to switch to using [`Markdown.tag`](https://github.com/JuliaLang/julia/blob/fafc0a4ee38d401aab2ac2027cc2b75c1d296faa/stdlib/Markdown/src/render/html.jl#L23) from stdlib?

**Comment by fonsp** [comment]
Hyperscript! https://github.com/yurivish/Hyperscript.jl/

**Comment by GiggleLiu** [comment]
> Hyperscript! https://github.com/yurivish/Hyperscript.jl/

I will have a look at this repo. Is this related to the

**Comment by fonsp** [comment]
Resolves #33

**Comment by GiggleLiu** [comment]
I read its README briefly. But not sure it fits this PR. If we change the HTML generator, we'd better change the other parts consistently. Shall I just update some doctest (so that we can check the correctness of examples in the docstrings) and get this PR merged first?

**Comment by fonsp** [comment]
I think we should rewrite this package to use Hyperscript for everything just to make it nicer to work with (and would have prevented some bugs that I missed), and Hyperscript has already solved [these problems](https://github.com/fonsp/PlutoUI.jl/pull/39#issuecomment-695788713) for us.

**Comment by GiggleLiu** [comment]
Sure, just feel free to either close this PR or merge it first. The Hyperscript thing should be in another PR.

File: src/Terminal.jl [hunk 2]
```diff
@@ -53,16 +53,18 @@ This allows you to to see the messages from `println`, `dump`, `Pkg.status`, etc
...
-    @warn "Oopsie!"
-end 
+```jldoctest
+julia> with_terminal() do
+           x=1+1
+           println(x)
+           @warn "Oopsie!"
+       end
+PlutoUI.WithTerminalOutput("2\\n┌ Warning: Oopsie!\\n└ @ Main none:4\\n", "", nothing)
 ```
```
Comment (fonsp) [inline src/Terminal.jl:63]: I'm happy that you took the initiative to write tests, but these `jldoctest` tests are actually unhelpful, because:
1. We don't want to test the `text/plain` representation of PlutoUI results
2. Having these `text/plain` outputs in the docs is confusing for a number of reasons, the main one that this is not what you get when you run it in Pluto.

**Review by fonsp — CHANGES_REQUESTED** [review]

**Comment by fonsp** [comment]
Let's do the Hyperscript rewrite instead? That also moves some of the testing burden to Hyperscript :)

File: src/Terminal.jl (hunk 2 above)
Comment (GiggleLiu) [reply src/Terminal.jl:63]: jldoctest are helpful in ensure the output correct. But agree text/plain does not help. I just haven't found a better solution.

**Comment by GiggleLiu** [comment]
Agree, then let's just close this PR.
